### PR TITLE
Use AES-256-CTR and SHA256

### PR DESCRIPTION
**Pull request by sebastian-meyer**
Currently Kitodo.Presentation uses a blowfish algorithm (`BF-CFB`) for encrypting/decrypting form data. This is considered weak nowadays (see [SonarCloud Analysis](https://sonarcloud.io/project/issues?id=kitodo-presentation&resolved=false&severities=BLOCKER%2CCRITICAL%2CMAJOR%2CMINOR&sinceLeakPeriod=true&types=VULNERABILITY)).

This pull request changes the defaults to `AES-256-CTR` as ciphering algorithm and `SHA256` as hashing algorithm.

Furthermore it removes the additional hashed value, because this opens a potential security hole and provides no real value.

### CODE REVIEW

File: Classes/Common/Helper.php [hunk 1]
```diff
@@ -214,26 +236,58 @@ public static function devLog($message, $severity = 0)
...
         if (empty($GLOBALS['TYPO3_CONF_VARS']['SYS']['encryptionKey'])) {
             self::devLog('No encryption key set in TYPO3 configuration', DEVLOG_SEVERITY_ERROR);
-            return;
+            return false;
+        }
+        // Generate random initialisation vector.
+        $iv = openssl_random_pseudo_bytes(openssl_cipher_iv_length(self::$cipherAlgorithm));
+        $key = openssl_digest($GLOBALS['TYPO3_CONF_VARS']['SYS']['encryptionKey'], self::$hashAlgorithm, true);
+        // Encrypt data.
+        $encrypted = openssl_encrypt($string, self::$cipherAlgorithm, $key, OPENSSL_RAW_DATA, $iv);
```
Comment (albig) [inline Classes/Common/Helper.php:286]: This fails with the following exception (including binary data):

<pre>
PHP Warning: openssl_encrypt(): A tag should be provided when using AEAD mode in /var/www/extensions/dlf/Classes/Common/Helper.php line 285

at openssl_encrypt('dlfCore1', 'aes-256-gcm', '��NÒ�&���|��m���UuV�P�|�|uG(@', 1, 'Ie�:UsZ�')
</pre>

What am I missing? I've just have the search plugin on the page.

**Review by albig — CHANGES_REQUESTED** [review]
I cannot test this.

Actually this is a lot of code to do some core functionality. Maybe we should better invest the time to swith to fluid templates?

File: Classes/Common/Helper.php (hunk 1 above)
Comment (sebastian-meyer) [reply Classes/Common/Helper.php:286]: Sorry, my fault! I experimented with different cipher algorithms and didn't push the final choice to upstream... This should be fixed now.

Explanation: The `*-gcm` and `*-ccm` cipher methods use authentication (AEAD mode) and therefore need a tag. But we don't need authentication, so I changed the cipher method to `AES-256-CTR`. This is the same as `AES-256-GCM`, but without authentication.

**Comment by sebastian-meyer** [comment]
> Actually this is a lot of code to do some core functionality. Maybe we should better invest the time to swith to fluid templates?

How would using fluid templates make this obsolete?

File: Classes/Common/Helper.php (hunk 1 above)
Comment (sebastian-meyer) [reply Classes/Common/Helper.php:286]: Although this worked fine, I added base64-encoding to get rid of the binary data in HTML. I think it's less confusing now.

**Comment by albig** [comment]
> > Actually this is a lot of code to do some core functionality. Maybe we should better invest the time to swith to fluid templates?
> 
> How would using fluid templates make this obsolete?

Protecting forms with hidden fields is daily business of the form-ViewHelper. Nothing you should think about on extension development.

**Review by albig — APPROVED** [review]
Works out-of-the box now. Thank you.

**Pull request merged**